### PR TITLE
Open62541 integration

### PR DESCRIPTION
**Pull request by nugatritter**
A wrapper template for the raw read service available as UA_Client_Service_read

### CODE REVIEW

**Comment by jalalmostafa** [comment]
Now there are two branches for this: @dirk-zimoch 's [paulscherrerinstitute:open62541-integration](https://github.com/paulscherrerinstitute/opcua/tree/open62541-integration) and @ralphlange 's [feature/open62541-integration](https://github.com/ralphlange/opcua/tree/feature/open62541-integration). I suggest we merge this pull request with feature/open62541-integration and open a new pull request for feature/open62541-integration to merge with ralphlange:master

**Comment by ralphlange** [comment]
I created the `feature/open62541-integration` branch in this repo (upstream fork) as a target for all Pull Requests within the scope of this activity.

Reason: The "normal" development of the Device Support will continue, this will not be a short activity, and I would like this addition to be somewhat robust and stable before merging into master. Merging now (and often) - at this point in the activity - would be likely to interfere with regular development.

So: please develop the open62541 integration in your personal forks - separate things could/should be done in separate branches.
Once you want more people to have a look at what you did or you think it is ready, create a Pull Request into `feature/open62541-integration` of the upstream fork. Once things are discussed and deemed fine, we'll merge it there.

Only when the whole combined effort of the open62541 integration (represented by the `feature/open62541-integration` branch of the upstream fork) is in a working and robust state, we should discuss to merge it into `master` to be released.

**Comment by ralphlange** [comment]
> We can add the compiler argument `USR_CXXFLAGS_WIN32+=-DNOMINMAX` to ~the Makefile~ configure/CONFIG_SITE of this project.

Sold. Perfect!

File: devOpcuaSup/open62541/DataElementOpen62541.h [hunk 1]
```diff
@@ -135,15 +135,15 @@ inline bool isWithinRange<UA_Int64, epicsUInt64> (const epicsUInt64 &value) {
...
+//}
 
-template<>
-inline bool isWithinRange<epicsInt64, UA_UInt64> (const UA_UInt64 &value) {
-    return !(value > static_cast<UA_UInt64>(std::numeric_limits<epicsInt64>::max()));
-}
+//template<>                                                                                                      // same as bool isWithinRange<UA_Uint64, epicsUInt64>(const epicsUInt64 &value)
+//inline bool isWithinRange<epicsInt64, UA_UInt64> (const UA_UInt64 &value) {
+//    return !(value > static_cast<UA_UInt64>(std::numeric_limits<epicsInt64>::max()));
+//}
```
Comment (dirk-zimoch) [inline devOpcuaSup/open62541/DataElementOpen62541.h:146]: It is not the same -- on 64 bit Linux at least. Removing those gives this warning with gcc 6.2.1
In file included from ../../devOpcuaSup/open62541/DataElementOpen62541.cpp:29:0:
../../devOpcuaSup/open62541/DataElementOpen62541.h: In instantiation of ‘bool DevOpcua::isWithinRange(const FROM&) [with TO = int; FROM = long unsigned int]’:
../../devOpcuaSup/open62541/DataElementOpen62541.h:838:61:   required from ‘long int DevOpcua::DataElementOpen62541::readScalar(ET*, dbCommon*, DevOpcua::ProcessReason*, epicsUInt32*, char*, epicsUInt32) [with ET = int; dbCommon = dbCommon; epicsUInt32 = unsigned int]’
../../devOpcuaSup/open62541/DataElementOpen62541.cpp:444:97:   required from here
../../devOpcuaSup/open62541/DataElementOpen62541.h:99:20: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     return !(value < std::numeric_limits<TO>::lowest() || value > std::numeric_limits<TO>::max());
              ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../devOpcuaSup/open62541/DataElementOpen62541.h: In instantiation of ‘bool DevOpcua::isWithinRange(const FROM&) [with TO = long long int; FROM = long unsigned int]’:
../../devOpcuaSup/open62541/DataElementOpen62541.h:838:61:   required from ‘long int DevOpcua::DataElementOpen62541::readScalar(ET*, dbCommon*, DevOpcua::ProcessReason*, epicsUInt32*, char*, epicsUInt32) [with ET = long long int; dbCommon = dbCommon; epicsUInt32 = unsigned int]’
../../devOpcuaSup/open62541/DataElementOpen62541.cpp:455:97:   required from here
../../devOpcuaSup/open62541/DataElementOpen62541.h:99:20: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

Comment (ralphlange) [reply devOpcuaSup/open62541/DataElementOpen62541.h:146]: Both the EPICS types and the UA types are explicit. There should be no ambiguity.

**Comment by nugatritter** [comment]
on Win64

"isWithinRange<epicsInt64, UA_UInt64>" is resolved into 
"isWithinRange<long long, unsigned long long>"

and

"isWithinRange<UA_Uint64, epicsUInt64> is resolved into 
"isWithinRange<long long, unsigned long long>"

so they are identical on Win64

but the signed / unsigned warning is obviously

Am 05.02.2021 um 10:24 schrieb Dirk Zimoch:

**Comment by ralphlange** [comment]
You are saying `UA_UInt64` is `unsigned long long` and `long long` at the same time?
I am not convinced.

**Comment by dirk-zimoch** [comment]
`UA_Uint64` should not be `long long` but `unsigned long long`.

**Comment by nugatritter** [comment]
UA_Uint64 is defined in open62541\include\open62541\types.h as typedef 
int64_t UA_Int64;

int64_t is defined in Microsoft Visual 
Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include\stdint.h as 
typedef long long int64_t;

Am 05.02.2021 um 10:51 schrieb Dirk Zimoch:

**Comment by nugatritter** [comment]
> 
> 
> You are saying `UA_UInt64` is `unsigned long long` and `long long` at the same time?
> I am not convinced.

**Comment by ralphlange** [comment]
Above, it's about `UA_UInt64`, though.

**Comment by ralphlange** [comment]
`Int` and `UInt` seem to have been mixed up at some point.

**Comment by nugatritter** [comment]
Whops, it's not closed, Sorry!

**Comment by dirk-zimoch** [comment]
> UA_Uint64 is defined in open62541\include\open62541\types.h as typedef
int64_t UA_Int64;

Does not match. You say `UA_Uint64` is defined but instead `UA_Int64` is defined.

**Comment by dirk-zimoch** [comment]
I have:
`typedef int64_t UA_Int64;`
and
`typedef uint64_t UA_UInt64;`

**Comment by dirk-zimoch** [comment]
EPICS has:
`typedef long long       epicsInt64;`
and
`typedef unsigned long long epicsUInt64;`
`

**Comment by dirk-zimoch** [comment]
stdint.h has for 64 bit Linux:
`typedef long int int64_t;`
and
`typedef unsigned long int uint64_t;`
but for 32 bit Linux:
`typedef long long int int64_t;`
and
`typedef unsigned long long int uint64_t;`

**Comment by ralphlange** [comment]
My brain hurts.

**Comment by dirk-zimoch** [comment]
On 64 bit Linux, `long` has 64 bits but on 64 bit Windows `long` has still only 32 bits because they did not want to break all their 32 bit code.

**Comment by nugatritter** [comment]
When I uncomment `bool isWithinRange<epicsInt64, UA_UInt64> (const UA_UInt64 &value)` in opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h
the compilers says:
cl -EHsc -GR               -nologo -FC -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE   -Ox -GL -Oy-   -W3 -w44355 -w44344 -w44251    -DUSE_TYPED_RSET    -MD -DEPICS_BUILD_DLL -DEPICS_CALL_DLL -TP  -I. -I../O.Common -I. -I. -I.. -I../../devOpcuaSup -I../../devOpcuaSup/open62541 -I../../include/compiler/msvc -I../../include/os/WIN32 -I../../include   -ID:\Profile\ggf\source\repos\EPICS\gtest/include -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/compiler/msvc -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/os/WIN32 -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include    -ID:\Profile\ggf\source\repos\open62541/include -ID:\Profile\ggf\source\repos\open62541/plugins/include -ID:\Profile\ggf\source\repos\open62541/arch -ID:\Profile\ggf\source\repos\open62541/build/src_generated -I$(OPENSSL)/include/openssl -I$(OPENSSL)/include/crypto    -c ../../devOpcuaSup/open62541/SessionOpen62541.cpp
SessionOpen62541.cpp
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(146): error C2766: Explizite Spezialisierung; "bool DevOpcua::isWithinRange<UA_Int64,epicsUInt64>(const epicsUInt64 &)" wurde bereits definiert.
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(134): note: Siehe vorherige Definition von "isWithinRange"
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(207): warning C4244: "=": Konvertierung von "__int64" in "epicsUInt32", möglicher Datenverlust
m

**Comment by nugatritter** [comment]
But when I comment it, the warning C4244 is there:
cl -EHsc -GR               -nologo -FC -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE   -Ox -GL -Oy-   -W3 -w44355 -w44344 -w44251    -DUSE_TYPED_RSET    -MD -DEPICS_BUILD_DLL -DEPICS_CALL_DLL -TP  -I. -I../O.Common -I. -I. -I.. -I../../devOpcuaSup -I../../devOpcuaSup/open62541 -I../../include/compiler/msvc -I../../include/os/WIN32 -I../../include   -ID:\Profile\ggf\source\repos\EPICS\gtest/include -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/compiler/msvc -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/os/WIN32 -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include    -ID:\Profile\ggf\source\repos\open62541/include -ID:\Profile\ggf\source\repos\open62541/plugins/include -ID:\Profile\ggf\source\repos\open62541/arch -ID:\Profile\ggf\source\repos\open62541/build/src_generated -I$(OPENSSL)/include/openssl -I$(OPENSSL)/include/crypto    -c ../../devOpcuaSup/open62541/SessionOpen62541.cpp
SessionOpen62541.cpp
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(207): warning C4244: "=": Konvertierung von "__int64" in "epicsUInt32", möglicher Datenverlust

**Comment by nugatritter** [comment]
I'm afraid, we need different solutions for Linux and Windows here.

**Comment by dirk-zimoch** [comment]
I think using the native types `long`, `unsigned long`, `long long` and `unsigned long long` in the templates instead of UA or EPICS types should fixt it.

**Comment by nugatritter** [comment]
I agree

**Comment by nugatritter** [comment]
the warning C4244 can be solved via
DataElementOpen62541.h(207): `value = (epicsUInt32)v;`

**Comment by ralphlange** [comment]
Using a `static_cast<>()`, please - else you get the next warning.

**Comment by nugatritter** [comment]
The problem seems to be bigger :-(
The full make shows this:
cl -EHsc -GR               -nologo -FC -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE   -Ox -GL -Oy-   -W3 -w44355 -w44344 -w44251    -DUSE_TYPED_RSET    -MD -DEPICS_BUILD_DLL -DEPICS_CALL_DLL -TP  -I. -I../O.Common -I. -I. -I.. -I../../devOpcuaSup -I../../devOpcuaSup/open62541 -I../../include/compiler/msvc -I../../include/os/WIN32 -I../../include   -ID:\Profile\ggf\source\repos\EPICS\gtest/include -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/compiler/msvc -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/os/WIN32 -ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include    -ID:\Profile\ggf\source\repos\open62541/include -ID:\Profile\ggf\source\repos\open62541/plugins/include -ID:\Profile\ggf\source\repos\open62541/arch -ID:\Profile\ggf\source\repos\open62541/build/src_generated -I$(OPENSSL)/include/openssl -I$(OPENSSL)/include/crypto    -c ../../devOpcuaSup/open62541/DataElementOpen62541.cpp
DataElementOpen62541.cpp
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(833): warning C4244: "=": Konvertierung von "UA_Int64" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(444): note: Siehe Verweis auf die gerade kompilierte Instanziierung "long DevOpcua::DataElementOpen62541::readScalar<epicsInt32>(ET *,dbCommon *,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der Funktions-Vorlage.
        with
        [
            ET=epicsInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(839): warning C4244: "=": Konvertierung von "UA_UInt64" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(845): warning C4244: "=": Konvertierung von "UA_Float" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(851): warning C4244: "=": Konvertierung von "UA_Double" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(845): warning C4244: "=": Konvertierung von "UA_Float" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsInt64
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(455): note: Siehe Verweis auf die gerade kompilierte Instanziierung "long DevOpcua::DataElementOpen62541::readScalar<epicsInt64>(ET *,dbCommon *,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der Funktions-Vorlage.
        with
        [
            ET=epicsInt64
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(851): warning C4244: "=": Konvertierung von "UA_Double" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsInt64
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(833): warning C4244: "=": Konvertierung von "UA_Int64" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsUInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(466): note: Siehe Verweis auf die gerade kompilierte Instanziierung "long DevOpcua::DataElementOpen62541::readScalar<epicsUInt32>(ET *,dbCommon *,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der Funktions-Vorlage.
        with
        [
            ET=epicsUInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(839): warning C4244: "=": Konvertierung von "UA_UInt64" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsUInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(845): warning C4244: "=": Konvertierung von "UA_Float" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsUInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(851): warning C4244: "=": Konvertierung von "UA_Double" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsUInt32
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(833): warning C4244: "=": Konvertierung von "UA_Int64" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsFloat64
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(477): note: Siehe Verweis auf die gerade kompilierte Instanziierung "long DevOpcua::DataElementOpen62541::readScalar<epicsFloat64>(ET *,dbCommon *,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der Funktions-Vorlage.
        with
        [
            ET=epicsFloat64
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(839): warning C4244: "=": Konvertierung von "UA_UInt64" in "ET", möglicher Datenverlust
        with
        [
            ET=epicsFloat64
        ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(953): warning C4267: "=": Konvertierung von "size_t" nach "epicsUInt32", Datenverlust möglich
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(675): note: Siehe Verweis auf die gerade kompilierte Instanziierung "long DevOpcua::DataElementOpen62541::readArray<epicsInt8>(ET *,const epicsUInt32,epicsUInt32 *,const UA_DataType *,dbCommon *,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der Funktions-Vorlage.
        with
        [
            ET=epicsInt8
        ]

**Comment by ralphlange** [comment]
In general, I agree with using the native types.

I still don't see why this problem does not appear in the UA SDK case, where the types should be the same, just having different names.

**Comment by ralphlange** [comment]
The `size_t` is another real difference between Linux and Windows.
Can be fixed using `static_cast<>()` (analog to dca492f on the `fix/windows-build` branch).

**Comment by dirk-zimoch** [comment]
Please check if My changes fix the problem. Biggest issue was long long to unsigned long because on some systems long and long long  are the same and on others not.

**Review by nugatritter — COMMENTED** [review]
Where is MAXHOSTNAMELEN defined?

**Comment by nugatritter** [comment]
still there

cl -EHsc -GR               -nologo -FC -D__STDC__=0 
-D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE   -Ox -GL -Oy-   
-W3 -w44355 -w44344 -w44251    -DUSE_TYPED_RSET -DNOMINMAX    -MD 
-DEPICS_BUILD_DLL -DEPICS_CALL_DLL -TP  -I. -I../O.Common -I. -I. -I.. 
-I../../devOpcuaSup -I../../devOpcuaSup/open62541 
-I../../include/compiler/msvc -I../../include/os/WIN32 -I../../include 
-ID:\Profile\ggf\source\repos\EPICS\gtest/include 
-ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/compiler/msvc 
-ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include/os/WIN32 
-ID:\Profile\ggf\source\repos\EPICS\base\3.15.8/include 
-ID:\Profile\ggf\source\repos\open62541/include 
-ID:\Profile\ggf\source\repos\open62541/plugins/include 
-ID:\Profile\ggf\source\repos\open62541/arch 
-ID:\Profile\ggf\source\repos\open62541/build/src_generated    -c 
../../devOpcuaSup/open62541/DataElementOpen62541.cpp
DataElementOpen62541.cpp
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(217): 
warning C4244: "=": Konvertierung von "__int64" in "epicsUInt32", 
möglicher Datenverlust
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(843): 
warning C4244: "=": Konvertierung von "UA_Int64" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(439): 
note: Siehe Verweis auf die gerade kompilierte Instanziierung "long 
DevOpcua::DataElementOpen62541::readScalar<epicsInt32>(ET *,dbCommon 
*,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der 
Funktions-Vorlage.
         with
         [
             ET=epicsInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(849): 
warning C4244: "=": Konvertierung von "UA_UInt64" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(855): 
warning C4244: "=": Konvertierung von "UA_Float" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(861): 
warning C4244: "=": Konvertierung von "UA_Double" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(855): 
warning C4244: "=": Konvertierung von "UA_Float" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsInt64
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(450): 
note: Siehe Verweis auf die gerade kompilierte Instanziierung "long 
DevOpcua::DataElementOpen62541::readScalar<epicsInt64>(ET *,dbCommon 
*,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der 
Funktions-Vorlage.
         with
         [
             ET=epicsInt64
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(861): 
warning C4244: "=": Konvertierung von "UA_Double" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsInt64
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(843): 
warning C4244: "=": Konvertierung von "UA_Int64" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsUInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(461): 
note: Siehe Verweis auf die gerade kompilierte Instanziierung "long 
DevOpcua::DataElementOpen62541::readScalar<epicsUInt32>(ET *,dbCommon 
*,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der 
Funktions-Vorlage.
         with
         [
             ET=epicsUInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(849): 
warning C4244: "=": Konvertierung von "UA_UInt64" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsUInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(855): 
warning C4244: "=": Konvertierung von "UA_Float" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsUInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(861): 
warning C4244: "=": Konvertierung von "UA_Double" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsUInt32
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(843): 
warning C4244: "=": Konvertierung von "UA_Int64" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsFloat64
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(472): 
note: Siehe Verweis auf die gerade kompilierte Instanziierung "long 
DevOpcua::DataElementOpen62541::readScalar<epicsFloat64>(ET *,dbCommon 
*,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der 
Funktions-Vorlage.
         with
         [
             ET=epicsFloat64
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(849): 
warning C4244: "=": Konvertierung von "UA_UInt64" in "ET", möglicher 
Datenverlust
         with
         [
             ET=epicsFloat64
         ]
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.h(963): 
warning C4267: "=": Konvertierung von "size_t" nach "epicsUInt32", 
Datenverlust möglich
D:\Profile\ggf\source\repos\EPICS\opcua_open\devOpcuaSup\open62541\DataElementOpen62541.cpp(671): 
note: Siehe Verweis auf die gerade kompilierte Instanziierung "long 
DevOpcua::DataElementOpen62541::readArray<epicsInt8>(ET *,const 
epicsUInt32,epicsUInt32 *,const UA_DataType *,dbCommon 
*,DevOpcua::ProcessReason *,epicsUInt32 *,char *,const epicsUInt32)" der 
Funktions-Vorlage.
         with
         [
             ET=epicsInt8
         ]

Am 05.02.2021 um 15:00 schrieb Dirk Zimoch:
>
> Please check if my changes fix the problem. Biggest issue was long 
> long to unsigned long because on some systems long and long long are 
> the same and on others not.
>

**Comment by dirk-zimoch** [comment]
> Where is MAXHOSTNAMELEN defined?

Stackoverflow asys: use MAXHOSTNAMELEN  instead of HOST_NAME_MAX. Does it not help for Windows?

**Comment by nugatritter** [comment]
No, unknown define.

**Comment by nugatritter** [comment]
When I build it with VS2019 the warning C4244 doesn't appear. Is this a good or bad result 😬

**Comment by nugatritter** [comment]
okay, the -W1 for compiler was the reason
with -W3 I see the warnings again (and a lot more 😞 )

**Review by nugatritter — COMMENTED** [review]
build looks good now
have a nice weekend!

**Pull request merged**